nixos/modules/services/hardware/udev.nix

```diff
@@ -298,16 +334,54 @@ in
 
     boot.kernelParams = mkIf (!config.networking.usePredictableInterfaceNames) [ "net.ifnames=0" ];
 
-    boot.initrd.extraUdevRulesCommands = optionalString (cfg.initrdRules != "")
+    boot.initrd.extraUdevRulesCommands = optionalString (!config.boot.initrd.systemd.enable && config.boot.initrd.services.udev.rules != "")
       ''
         cat <<'EOF' > $out/99-local.rules
-        ${cfg.initrdRules}
+        ${config.boot.initrd.services.udev.rules}
         EOF
       '';
 
+    boot.initrd.systemd.additionalUpstreamUnits = [
+      # TODO: "initrd-udevadm-cleanup-db.service" is commented out because of https://github.com/systemd/systemd/issues/12953
+      "systemd-udevd-control.socket"
+      "systemd-udevd-kernel.socket"
+      "systemd-udevd.service"
+      "systemd-udev-settle.service"
+      "systemd-udev-trigger.service"
+    ];
+    boot.initrd.systemd.storePaths = [
+      "${config.boot.initrd.systemd.package}/lib/systemd/systemd-udevd"
+      "${config.boot.initrd.systemd.package}/lib/udev"
+    ] ++ map toString config.boot.initrd.services.udev.binPackages;
```

Comment:
Shouldn't this append `"/bin"` to the packages?

Reply:
usually the binaries require more files than just the binary in the initrd; so this is  'safer' to do I suppose than to blindly append `/bin`; for a slightly larger initrd. Is that correct @dasJ ?

Reply:
It's safer but I don't know if it may be worth it to only add the binaries. There is so much stuff like man pages we don't need

Reply:
What could be needed other than the binaries for this? The only thing we use `binPackages` for is to fix udev rules to refer to programs by absolute path, isn't it? Heck, ideally, we'd have a way to keep track of which binaries we actually fix up this way so we could add only those ones.